src/Npgsql/TypeHandlers/NumericHandlers/NumericHandler.cs

```diff
@@ -97,7 +97,13 @@ public override decimal Read(NpgsqlReadBuffer buf, int len, FieldDescription fie
             if (sign == 0xC000)
                 throw new NpgsqlSafeReadException(new InvalidCastException("Numeric NaN not supported by System.Decimal"));
 
-            result = Math.Round(result, dscale);
+            var rscale = ((decimal.GetBits(result)[3] >> 16) & 0x7F);                           //Find scale of result
+
+            if (rscale < dscale)                                                                //If result scale is insufficient,
+                result = result * new decimal((int)Math.Pow(10, dscale), 0, 0, false, 2);       //increase it.
```

Comment:
So in the end we don't have to look at the type modifier to know what scale to round to?

Reply:
That is right. My original understanding that `dscale` was the scale of the data rather than the scale of the `column` was incorrect.

Reply:
`Math.Pow` is a perf killer. Use a table of predefined values (powers of 10).

```csharp
static readonly decimal[] Powers10 =
{
    1M,
    10M,
    100M,
    1000M,
    10000M,
    100000M,
    1000000M,
    10000000M,
    100000000M,
    1000000000M,
    10000000000M,
    // And so on. You should totally have 29 powers of 10.
};
```

```csharp
result.Value *= new DecimalRaw { Value = Powers10[scale - resultScale], Flags = (scale - resultScale) << DecimalRaw.ScaleShift }.Value;
```

It's okay to have 1000000000 as the max value in the table and to store `int`s in it, but then you should implement a method to produce a `DecimalRaw` value with the required scale.